lib/Dav/PropFindPlugin.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright 2024 Christopher Ng <chrng8@gmail.com>
+ *
+ * @author Christopher Ng <chrng8@gmail.com>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Files_DownloadLimit\Dav;
+
+use OCA\DAV\Connector\Sabre\File as SabreFile;
+use OCA\Files_DownloadLimit\Db\Limit;
+use OCA\Files_DownloadLimit\Db\LimitMapper;
+use OCA\Files_DownloadLimit\LimitList;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\Files\File;
+use OCP\IUser;
+use OCP\IUserSession;
+use OCP\Share\IManager as IShareManager;
+use OCP\Share\IShare;
+use Psr\Log\LoggerInterface;
+use Sabre\DAV\INode;
+use Sabre\DAV\PropFind;
+use Sabre\DAV\Server;
+use Sabre\DAV\ServerPlugin;
+
+class PropFindPlugin extends ServerPlugin {
+	public const DOWNLOAD_LIMITS_PROPERTY = '{http://nextcloud.org/ns}share-download-limits';
+
+	public function __construct(
+		private IShareManager $shareManager,
+		private IUserSession $userSession,
+		private LimitMapper $limitMapper,
+		private LoggerInterface $logger,
+	) {
+	}
+
+	public function initialize(Server $server): void {
+		$server->on('propFind', [$this, 'propFind']);
+	}
+
+	public function propFind(PropFind $propFind, INode $node) {
+		if (!in_array(static::DOWNLOAD_LIMITS_PROPERTY, $propFind->getRequestedProperties(), true)) {
+			return;
+		}
+
+		if (!($node instanceof SabreFile)) { // Only allowed on files
+			return;
+		}
+
+		$propFind->handle(
+			static::DOWNLOAD_LIMITS_PROPERTY,
+			function () use ($node) {
+				$user = $this->userSession->getUser();
+				if (!($user instanceof IUser)) {
+					return new LimitList([]);
+				}
+
+				$externalShares = $this->getSharesForTypes(
```

Comment:
Aren't we in the context of a specific share here? So the limit should be relative to the share and not every share.
Or maybe I missed something.

Reply:
The context is a single file node which can have multiple external shares

Reply:
So the answer is for the share owner when he edits the share?

Reply:
This would be for clients I guess as requested in this issue https://github.com/nextcloud/files_downloadlimit/issues/239, no plan to use this for the web UI

Reply:
Ok, not sure what is the purpose then :thinking:

Reply:
Any more details on clients-side @tobiasKaminsky?